ui/library/src/utils/VisibilityDetector.ts

```diff
@@ -0,0 +1,69 @@
+export type SetEntriesCallback<TEntry> = (visible: Set<TEntry>) => void;
+export type OnVisibilityChangeCallback<TEntry> = (args: {
+  visibleElements: Element[];
+  hiddenElements: Element[];
+  lastEntries: Set<TEntry>;
+}) => Set<TEntry>;
+
+export default class VisibilityDetector<TEntry> {
+  _root: Element;
+  _observer: IntersectionObserver;
+  _lastEntries: Set<TEntry>;
+  _setEntries: SetEntriesCallback<TEntry>;
+  _onVisibilityChange: OnVisibilityChangeCallback<TEntry>;
+
+  constructor({
+    root,
+    setEntries,
+    onVisibilityChange,
+  }: {
+    root: Element;
+    setEntries: SetEntriesCallback<TEntry>;
+    onVisibilityChange: OnVisibilityChangeCallback<TEntry>;
+  }) {
+    this._root = root;
+    this._lastEntries = new Set();
+    this._setEntries = setEntries;
+    this._onVisibilityChange = onVisibilityChange;
+    this._observer = new IntersectionObserver(
+      entries => {
+        // Collect visible and hidden elements
+        const visibleElements: Element[] = [];
+        const hiddenElements: Element[] = [];
+        for (const entry of entries) {
+          if (entry.isIntersecting) {
+            visibleElements.push(entry.target);
+          } else {
+            hiddenElements.push(entry.target);
+          }
+        }
```

Comment:
The [type of the entries](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver/IntersectionObserver) in the **IntersectionObserver** callback function will be `IntersectionObserverEntry`. So I think it is incorrect to put those entries in variables that have `Element[]` type. Instead, it's better to change the parameter types in `OnVisibilityChangeCallback` to `IntersectionObserverEntry` (see my comments above at line 4).

Also, after you make the change, you can write like below to make these two lines more elegant
```
const visibleEntries = entries.filter(entry => entry.isIntersecting)
const hiddenEntries = entries.filter(entry => !entry.isIntersecting)
```